File: typescript/tests/test_streaming.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { Azex } from "../src/client.js";
import { Stream } from "../src/streaming.js";

function makeSseBody(lines: string[]): ReadableStream<Uint8Array> {
  const encoder = new TextEncoder();
  const text = lines.join("\n") + "\n";
  return new ReadableStream({
    start(controller) {
      controller.enqueue(encoder.encode(text));
      controller.close();
    },
  });
}

function makeOpenAiSseChunks(contents: string[]): string[] {
  const lines: string[] = [];
  for (const content of contents) {
    const chunk = {
      id: "chatcmpl-123",
      object: "chat.completion.chunk",
      created: 1711234567,
      model: "openai/gpt-4o",
      choices: [{ index: 0, delta: { content }, finish_reason: null }],
    };
    lines.push(`data: ${JSON.stringify(chunk)}`);
    lines.push("");
  }
  lines.push("data: [DONE]");
  lines.push("");
  return lines;
}

function makeAnthropicSseEvents(): string[] {
  const events = [
    {
      type: "message_start",
      message: {
        id: "msg_123",
        type: "message",
        role: "assistant",
        content: [],
        model: "anthropic/claude-sonnet-4-6",
        stop_reason: null,
        stop_sequence: null,
        usage: { input_tokens: 10, output_tokens: 0 },
      },
    },
    {
      type: "content_block_start",
      index: 0,
      content_block: { type: "text", text: "" },
    },
    {
      type: "content_block_delta",
      index: 0,
      delta: { type: "text_delta", text: "Hello" },
    },
    {
      type: "content_block_delta",
      index: 0,
      delta: { type: "text_delta", text: " world" },
    },
    { type: "content_block_stop", index: 0 },
    {
      type: "message_delta",
      delta: { stop_reason: "end_turn" },
      usage: { output_tokens: 2 },
    },
    { type: "message_stop" },
  ];

  const lines: string[] = [];
  for (const event of events) {
    lines.push(`data: ${JSON.stringify(event)}`);
    lines.push("");
  }
  return lines;
}

describe("Stream (OpenAI SSE format)", () => {
  let client: Azex;
  let fetchSpy: ReturnType<typeof vi.spyOn>;

  beforeEach(() => {
    client = new Azex({ apiKey: "sk_test_1234", maxRetries: 0 });
    fetchSpy = vi.spyOn(globalThis, "fetch");
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("yields chunks from OpenAI SSE stream", async () => {
    const sseLines = makeOpenAiSseChunks(["Hello", " world", "!"]);
    fetchSpy.mockResolvedValueOnce(
      new Response(makeSseBody(sseLines), {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      }),
    );

    const stream = await client.chat.completions.stream({
      model: "openai/gpt-4o",
      messages: [{ role: "user", content: "Hi" }],
    });

    const collected: string[] = [];
    for await (const chunk of stream) {
      const content = chunk.choices[0]?.delta?.content;
      if (content) collected.push(content);
    }

    expect(collected).toEqual(["Hello", " world", "!"]);
  });

  it("emits text events via .on('text', ...)", async () => {
    const sseLines = makeOpenAiSseChunks(["Hi", " there"]);
    fetchSpy.mockResolvedValueOnce(
      new Response(makeSseBody(sseLines), {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      }),
    );

    const stream = await client.chat.completions.stream({
      model: "openai/gpt-4o",
      messages: [{ role: "user", content: "Hello" }],
    });

    const texts: string[] = [];
    stream.on("text", (t) => texts.push(t));

    // consume the stream
    for await (const _ of stream) { /* drain */ }

    expect(texts).toEqual(["Hi", " there"]);
  });

  it("getFinalCompletion returns last chunk", async () => {
    const sseLines = makeOpenAiSseChunks(["A", "B"]);
    fetchSpy.mockResolvedValueOnce(
      new Response(makeSseBody(sseLines), {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      }),
    );

    const stream = await client.chat.completions.stream({
      model: "openai/gpt-4o",
      messages: [{ role: "user", content: "Hi" }],
    });

    for await (const _ of stream) { /* drain */ }

    const final = stream.getFinalCompletion();
    expect(final).toBeDefined();
    expect(final?.choices[0]?.delta?.content).toBe("B");
  });
});

describe("Stream (Anthropic SSE format)", () => {
  let client: Azex;
  let fetchSpy: ReturnType<typeof vi.spyOn>;

  beforeEach(() => {
    client = new Azex({ apiKey: "sk_test_1234", maxRetries: 0 });
    fetchSpy = vi.spyOn(globalThis, "fetch");
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("yields Anthropic SSE events", async () => {
    const sseLines = makeAnthropicSseEvents();
    fetchSpy.mockResolvedValueOnce(
      new Response(makeSseBody(sseLines), {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      }),
    );

    const stream = await client.messages.stream({
      model: "anthropic/claude-sonnet-4-6",
      messages: [{ role: "user", content: "Hi" }],
      max_tokens: 100,
    });

    const events: string[] = [];
    for await (const event of stream) {
      events.push(event.type);
    }

    expect(events).toContain("message_start");
    expect(events).toContain("content_block_delta");
    expect(events).toContain("message_stop");
  });

  it("emits text from Anthropic content_block_delta events", async () => {
    const sseLines = makeAnthropicSseEvents();
    fetchSpy.mockResolvedValueOnce(
      new Response(makeSseBody(sseLines), {
        status: 200,
        headers: { "Content-Type": "text/event-stream" },
      }),
    );

    const stream = await client.messages.stream({
      model: "anthropic/claude-sonnet-4-6",
      messages: [{ role: "user", content: "Hi" }],
      max_tokens: 100,
    });

    const texts: string[] = [];
    stream.on("text", (t) => texts.push(t));
    for await (const _ of stream) { /* drain */ }

    expect(texts).toEqual(["Hello", " world"]);
  });

  it("Stream implements AsyncIterable", () => {
    const response = new Response(makeSseBody([]), {
      status: 200,
      headers: { "Content-Type": "text/event-stream" },
    });
    const stream = new Stream(response, (data) => {
      try {
        return JSON.parse(data);
      } catch {
        return null;
      }
    });
    expect(stream[Symbol.asyncIterator]).toBeDefined();
    expect(typeof stream[Symbol.asyncIterator]).toBe("function");
  });
});

File: python/tests/test_streaming.py
"""Tests for SSE streaming (OpenAI and Anthropic formats)."""

from __future__ import annotations

import json
from typing import List

import pytest
import respx
import httpx

from azex import Azex
from azex._streaming import Stream, make_chunk_parser


def _make_openai_sse(chunks: list) -> bytes:
    lines = []
    for chunk in chunks:
        lines.append(f"data: {json.dumps(chunk)}")
        lines.append("")
    lines.append("data: [DONE]")
    lines.append("")
    return "\n".join(lines).encode()


def _make_anthropic_sse(events: list) -> bytes:
    lines = []
    for event in events:
        lines.append(f"event: {event['type']}")
        lines.append(f"data: {json.dumps(event)}")
        lines.append("")
    return "\n".join(lines).encode()


OPENAI_CHUNKS = [
    {
        "id": "chatcmpl-1",
        "object": "chat.completion.chunk",
        "created": 1700000000,
        "model": "openai/gpt-4o",
        "choices": [{"index": 0, "delta": {"role": "assistant", "content": "The"}, "finish_reason": None}],
    },
    {
        "id": "chatcmpl-1",
        "object": "chat.completion.chunk",
        "created": 1700000000,
        "model": "openai/gpt-4o",
        "choices": [{"index": 0, "delta": {"content": " answer"}, "finish_reason": None}],
    },
    {
        "id": "chatcmpl-1",
        "object": "chat.completion.chunk",
        "created": 1700000000,
        "model": "openai/gpt-4o",
        "choices": [{"index": 0, "delta": {"content": " is 42."}, "finish_reason": "stop"}],
    },
]

ANTHROPIC_EVENTS = [
    {"type": "message_start", "message": {
        "id": "msg_abc",
        "type": "message",
        "role": "assistant",
        "content": [],
        "model": "anthropic/claude-sonnet-4-6",
        "stop_reason": None,
        "stop_sequence": None,
        "usage": {"input_tokens": 10, "output_tokens": 0},
    }},
    {"type": "content_block_start", "index": 0, "content_block": {"type": "text", "text": ""}},
    {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "Hello "}},
    {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "world!"}},
    {"type": "content_block_stop", "index": 0},
    {"type": "message_delta", "delta": {"stop_reason": "end_turn"}, "usage": {"input_tokens": 10, "output_tokens": 5}},
    {"type": "message_stop"},
]


@respx.mock
def test_openai_stream_yields_all_chunks() -> None:
    respx.post("https://api.azex.ai/v1/chat/completions").mock(
        return_value=httpx.Response(
            200,
            content=_make_openai_sse(OPENAI_CHUNKS),
            headers={"content-type": "text/event-stream"},
        )
    )
    client = Azex(api_key="test-key")
    stream = client.chat.completions.create(
        model="openai/gpt-4o",
        messages=[{"role": "user", "content": "What is the answer?"}],
        stream=True,
    )
    chunks = list(stream)
    assert len(chunks) == 3
    texts = [c["choices"][0]["delta"].get("content", "") for c in chunks]
    assert "".join(t for t in texts if t) == "The answer is 42."


@respx.mock
def test_stream_get_final_completion() -> None:
    respx.post("https://api.azex.ai/v1/chat/completions").mock(
        return_value=httpx.Response(
            200,
            content=_make_openai_sse(OPENAI_CHUNKS),
            headers={"content-type": "text/event-stream"},
        )
    )
    client = Azex(api_key="test-key")
    stream = client.chat.completions.create(
        model="openai/gpt-4o",
        messages=[{"role": "user", "content": "hi"}],
        stream=True,
    )
    # consume
    list(stream)
    final = stream.get_final_completion()
    assert final is not None
    assert final["choices"][0]["delta"]["content"] == " is 42."


@respx.mock
def test_anthropic_stream_parses_event_types() -> None:
    respx.post("https://api.azex.ai/v1/messages").mock(
        return_value=httpx.Response(
            200,
            content=_make_anthropic_sse(ANTHROPIC_EVENTS),
            headers={"content-type": "text/event-stream"},
        )
    )
    client = Azex(api_key="test-key")
    stream = client.messages.create(
        model="anthropic/claude-sonnet-4-6",
        messages=[{"role": "user", "content": "hi"}],
        max_tokens=100,
        stream=True,
    )
    events = list(stream)
    types = [e.get("type") for e in events if isinstance(e, dict)]
    assert "content_block_delta" in types
    # Verify text was captured
    text_deltas = [
        e["delta"]["text"]
        for e in events
        if isinstance(e, dict) and e.get("type") == "content_block_delta"
    ]
    assert "".join(text_deltas) == "Hello world!"


def test_stream_cannot_be_iterated_twice() -> None:
    """Consuming a stream twice should raise RuntimeError."""
    parse = make_chunk_parser()
    # Build a minimal response
    body = b"data: {\"test\": 1}\n\ndata: [DONE]\n\n"
    response = httpx.Response(200, content=body, headers={"content-type": "text/event-stream"})
    stream: Stream = Stream(response, parse)
    list(stream)  # first consumption
    with pytest.raises(RuntimeError, match="already been consumed"):
        list(stream)


@respx.mock
def test_stream_context_manager() -> None:
    respx.post("https://api.azex.ai/v1/chat/completions").mock(
        return_value=httpx.Response(
            200,
            content=_make_openai_sse(OPENAI_CHUNKS),
            headers={"content-type": "text/event-stream"},
        )
    )
    client = Azex(api_key="test-key")
    raw_stream = client.chat.completions.create(
        model="openai/gpt-4o",
        messages=[{"role": "user", "content": "hi"}],
        stream=True,
    )
    with raw_stream as stream:
        chunks = list(stream)
    assert len(chunks) == 3
